nm dns: Protect global DNS config in checkpoint

cathay4t opened the pull request:
[NetworkManager has introduced][1] new flag for placing global DNS into
protection of checkpoint:
`NM_CHECKPOINT_CREATE_FLAG_TRACK_INTERNAL_GLOBAL_DNS: 0x20`.

Since nmstate are supposed to rollback on global DNS changes also,
nmstatectl is patched to create checkpoint with that flag first and
fallback to normal flag without it.

Integration test case included.
Kubernetes-nmstate depends on checkpoint to rollback to original state when
check fails, hence marked as tier1 test case.


Resolves: https://issues.redhat.com/browse/RHEL-29001

[1]: https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/1875

cathay4t commented:
To test this, you need to use `NetworkManager-1.47.2-33136.copr` or later version from copr repo of NetworkManager-main: https://copr.fedorainfracloud.org/coprs/networkmanager/NetworkManager-main/

ffmancera commented on tests/integration/nm/dns_test.py:
```diff
@@ -207,3 +210,50 @@ def test_global_dns_with_dns_options():
+
+
+@pytest.mark.skipif(
+    nm_minor_version() <= 40,
```
This should be 47, shouldn't it?

cathay4t replied:
Yes. Fixed.

ihuguet commented:
What's exactly the reason of trying with the flag and then without it? Is it to support older versions of NetworkManager?

I think we need a generic way to take certain actions depending on the version of NetworkManager, and not just trying if it works. Otherwise the verification is less robust because we don't know if it was supposed to fail or not.

For example it's something that I also wanted to use for https://github.com/nmstate/nmstate/pull/2557. The warning message added there is only valid for NM >= 1.42, so that check should only be done in that case.

It would have an additional benefit too: it would be easier to clean up the code when we raise the minimum supported version of NM. For example, if we add here an `if nm_version > 1.47 {...} else {...}`, we can remove the unused code when we raise the minimum NM version to 1.48.

cathay4t commented:
> What's exactly the reason of trying with the flag and then without it? Is it to support older versions of NetworkManager?
> 
> I think we need a generic way to take certain actions depending on the version of NetworkManager, and not just trying if it works. Otherwise the verification is less robust because we don't know if it was supposed to fail or not.
> 
> For example it's something that I also wanted to use for #2557. The warning message added there is only valid for NM >= 1.42, so that check should only be done in that case.
> 
> It would have an additional benefit too: it would be easier to clean up the code when we raise the minimum supported version of NM. For example, if we add here an `if nm_version > 1.47 {...} else {...}`, we can remove the unused code when we raise the minimum NM version to 1.48.

Nmstate has no plan to increase minimum NM version for the seek of supporting RHEL 9.2 (NetworkManager 1.42).

And meanwhile, I will try to backport NM patch to 1.42 branch, so checking the version is not correct here.

In python world, they encourage try-except instead of check-then-use-correct. I think that is also correct for this case.

ihuguet commented:
Ok. I still think that knowing the NM vesion would be helpful in many situations, but in this case it works this way too, so leave it if you prefer it. Anyway, please add a comment explaining why it may fail.

cathay4t commented:
> Ok. I still think that knowing the NM vesion would be helpful in many situations, but in this case it works this way too, so leave it if you prefer it. Anyway, please add a comment explaining why it may fail.

Comment added explaining why version check will not helps and why we try to create checkpoint again when fails.

ihuguet commented on rust/src/lib/nm/nm_dbus/dbus.rs:
```diff
@@ -95,6 +92,28 @@ impl<'a> NmDbus<'a> {
+                // supported by NM 1.47+ and might be backported to other
+                // versions. There is no way to know whether it is supported or
+                // not by checking the NM version. Hence we try to create
+                // the checkpoint with this flag on second try.
```
with -> without

cathay4t replied:
Fixed. Thanks!

cathay4t committed:
nm dns: Protect global DNS config in checkpoint

[NetworkManager has introduced][1] new flag for placing global DNS into
protection of checkpoint:
`NM_CHECKPOINT_CREATE_FLAG_TRACK_INTERNAL_GLOBAL_DNS: 0x20`.

Since nmstate are supposed to rollback on global DNS changes also,
nmstatectl is patched to create checkpoint with that flag first and
fallback to normal flag without it.

Integration test case included.

Kubernetes-nmstate depends on checkpoint to rollback to original state when
check fails, hence marked as tier1 test case.

Resolves: https://issues.redhat.com/browse/RHEL-29001

[1]: https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/1875

Signed-off-by: Gris Ge <fge@redhat.com>
